File: C++/Baekjoon/암호_만들기.cpp
// https://www.acmicpc.net/problem/1759
// 1759.암호 만들기 : 2020.10.08

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// 생성된 암호가 모음 1개 이상, 자음 2개 이상을 포함하는지 검사
bool isRightCipher(const vector<char> & cipher){
    // 모음 개수 세기
    int cnt = 0;
    vector<char> vowel = {'a', 'e', 'i', 'o', 'u'};
    for(int i=0, idx = 0; i<vowel.size(); i++){
        for(int j=idx; j<cipher.size(); j++){
            if(cipher[j] == vowel[i]){
                cnt++;
                idx = j + 1;
            }
        }
    }
    return (cnt >= 1 && cipher.size() - cnt >= 2);
}

// BackTracking 기법을 사용하여 암호를 만든다
void Cipher(const vector<char> & alpha, vector<char> & cipher, const int & l, int s){
    if(cipher.size() == l) {
        if(isRightCipher(cipher)){
            for(int i=0; i<cipher.size(); i++){
                cout<<cipher[i];
            }cout<<'\n';
        }
    }
    // 추가해야 하는 문자열의 수와 남은 문자열의 수를 비교하여 암호를 만들지 못할 것 같으면 그만둔다
    if(l - cipher.size() > alpha.size() - s) return;

    for(int i=s; i<alpha.size(); i++){
        cipher.push_back(alpha[i]);
        Cipher(alpha, cipher, l, i+1);
        cipher.pop_back();
    }
}

int main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL);
    int l, c; cin>>l>>c;
    vector<char> alpha(c), cipher;
    for(int i=0; i<c; i++){
        cin>>alpha[i];
    }
    sort(alpha.begin(), alpha.end());

    Cipher(alpha, cipher, l, 0);
}

File: python/baekjoon/암호_만들기.py
# https://www.acmicpc.net/problem/1759
# 1759.암호 만들기 : 2020.10.08

import sys

l, c = map(int, sys.stdin.readline().split())
alpha = sys.stdin.readline().split()
alpha = sorted(alpha)

# 생성된 암호가 모음 1개 이상, 자음 2개 이상 포함하는지 검사
def isRightCipher(cipher):
    cnt = 0
    vowels = ['a','e','i','o','u']
    for vowel in vowels:
        if vowel in cipher:
            cnt += 1

    return cnt >= 1 and len(cipher) - cnt >= 2

# BackTracking 기법을 사용하여 암호를 생성한다
def createCipher(alpha, cipher, s):
    if len(cipher) == l:
        if isRightCipher(cipher):
            for ci in cipher:
                print(ci, end='')
            print()
            return

    # 추가해야 하는 문자 개수가 사용할 수 있는 문자 개수보다 많다면 함수를 끝낸다
    if len(alpha) - s < l - len(cipher):
        return

    for i in range(s, c):
        cipher.append(alpha[i])
        createCipher(alpha, cipher, i+1)
        cipher.pop()

createCipher(alpha, [], 0)